File: bindings/python/scp_sdk/identity.py
"""SCP Identity types.

Phase 4 PR 5 Agent B+C (#1549) collapsed :class:`Identity` into a
pure handle wrapper exposing safe metadata only. Every operation that
used to live on :class:`Identity` as a classmethod or instance method
is now a method on :class:`scp_sdk.SCP` (see ADR-048 and
``scp_sdk.scp.SCP``).

Call sites shape::

    from scp_sdk import SCP, Identity
    from scp_sdk.types import CustodyType

    with SCP() as scp:
        identity = await scp.identity_create(CustodyType.IN_MEMORY)
        # identity is an Identity wrapper — use identity.did / identity.custody_type
        rotated = await scp.identity_rotate_key(identity._raw_handle)

:class:`DIDDocument`, :class:`IdentityAttestation`, and
:class:`RevocationStatus` are pure data classes; they hold no SCP
reference.

See ``.docs/adrs/phase-3.md`` ADR-014 for the underlying API design
and ADR-048 for the façade consolidation rationale.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from scp_sdk.errors import ValidationError
from scp_sdk.types import CustodyType

if TYPE_CHECKING:
    from scp_sdk.scp import SCP


# ---------------------------------------------------------------------------
# DIDDocument
# ---------------------------------------------------------------------------


@dataclass
class DIDDocument:
    """A resolved DID Document.

    Mirrors the ``PyDIDDocument`` returned by
    :meth:`scp_sdk.SCP.identity_resolve`. Fields are extracted from the
    opaque bridge object into a pure-Python dataclass for ergonomic
    access.
    """

    #: The DID string this document describes.
    id: str

    #: Verification methods (list of dicts with ``id``, ``type``,
    #: ``controller``, ``public_key_multibase``).
    verification_methods: list[dict[str, Any]] = field(default_factory=list)

    #: Service entries (list of dicts with ``id``, ``type``,
    #: ``service_endpoint``).
    services: list[dict[str, Any]] = field(default_factory=list)

    #: Alternative identifiers for the DID subject.
    also_known_as: list[str] = field(default_factory=list)

    #: Authentication method references.
    authentication: list[str] = field(default_factory=list)

    #: Assertion method references.
    assertion_methods: list[str] = field(default_factory=list)


def _bridge_doc_to_dataclass(bridge_doc: Any) -> DIDDocument:
    """Convert a ``PyDIDDocument`` bridge object to a :class:`DIDDocument`."""
    return DIDDocument(
        id=bridge_doc.id,
        verification_methods=list(bridge_doc.verification_methods),
        services=list(bridge_doc.services),
        also_known_as=list(bridge_doc.also_known_as),
        authentication=list(bridge_doc.authentication),
        assertion_methods=list(bridge_doc.assertion_methods),
    )


# ---------------------------------------------------------------------------
# Identity
# ---------------------------------------------------------------------------


class Identity:
    """An SCP identity handle.

    Pure handle wrapper: stores the opaque ``PyIdentity`` raw handle
    produced by the bridge and exposes the safe metadata fields
    (``did``, ``custody_type``, ``agent_public_key_multibase``).

    Construct via :meth:`scp_sdk.SCP.identity_create`,
    :meth:`~scp_sdk.SCP.identity_load`,
    :meth:`~scp_sdk.SCP.identity_create_with_agent_key`, and friends —
    those methods return ``Identity`` instances already wrapped.

    All mutating/effectful operations (``rotate_key``, ``add_agent_key``,
    ``attest_device``, ``execute_recovery``, attestation lifecycle, etc.)
    live as methods on :class:`scp_sdk.SCP`. Pass ``identity._raw_handle``
    when the SCP-level method needs the opaque bridge handle.
    """

    __slots__ = ("_raw_handle",)

    def __init__(self, handle: Any) -> None:
        """Wrap a ``PyIdentity`` bridge handle.

        Users should not call this directly — use the ``scp.identity_*``
        factory methods.
        """
        self._raw_handle = handle

    @classmethod
    def _from_handle(cls, _scp: SCP | None, raw: Any) -> Identity:
        """Internal constructor used by :class:`scp_sdk.SCP` methods.

        The ``_scp`` parameter is accepted for call-site symmetry across
        handle types but not stored — :class:`Identity` holds no SCP
        reference.
        """
        return cls(raw)

    # -- Properties ----------------------------------------------------------

    @property
    def did(self) -> str:
        """The DID string for this identity (e.g. ``"did:dht:z6Mk..."``)."""
        return self._raw_handle.did

    @property
    def custody_type(self) -> CustodyType | str:
        """The custody type used for this identity.

        Returns a :class:`~scp_sdk.types.CustodyType` enum member when
        the value matches a known variant, otherwise the raw string.
        """
        raw = self._raw_handle.custody
        try:
            return CustodyType(raw)
        except ValueError:
            return raw

    @property
    def agent_public_key_multibase(self) -> str | None:
        """The multibase-encoded ``#agent`` verification method key.

        Returns ``None`` when this identity does not have an agent key
        (see :meth:`scp_sdk.SCP.identity_add_agent_key`). Present only
        when the underlying bridge handle exposes the accessor — older
        builds may return ``None`` for either reason.
        """
        return getattr(self._raw_handle, "agent_public_key_multibase", None)

    # -- Dunder methods ------------------------------------------------------

    def __repr__(self) -> str:
        return f"Identity(did={self.did!r}, custody_type={self.custody_type!r})"

    def __str__(self) -> str:
        return self.did


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _parse_finite_int(value: object, field_name: str) -> int:
    """Coerce *value* to a non-negative integer, raising :class:`ValidationError` on:

    - ``bool`` (since ``isinstance(True, int)`` is ``True`` silently)
    - ``str`` (cross-SDK divergence: TypeScript ``Number.isFinite`` rejects strings)
    - non-integer ``float`` (silent truncation: ``1.5`` → ``1``)
    - NaN/Infinity floats (not representable as int)
    - negative integers (Unix timestamps are non-negative)

    Whole-number floats (e.g. ``1700000000.0``) are accepted and coerced to
    ``int`` because JSON parsers commonly deserialize integer fields as floats,
    and lossless coercion is correct behavior at a deserialization boundary.

    Raises ``ValidationError`` with code ``SCP-VALID-7005`` ("Invalid
    field value") to match the TypeScript SDK's behavior — both bridges
    parse the same wire format and should surface the same error type
    and code so cross-language consumers can handle them uniformly.
    """
    if isinstance(value, bool):
        raise ValidationError(
            f"{field_name} must be a finite integer, got bool",
            "SCP-VALID-7005",
        )
    if isinstance(value, str):
        raise ValidationError(
            f"{field_name} must be a number, got str",
            "SCP-VALID-7005",
        )
    if not isinstance(value, (int, float)):
        raise ValidationError(
            f"{field_name} must be a number, got {type(value).__name__}",
            "SCP-VALID-7005",
        )
    if isinstance(value, float):
        import math

        if not math.isfinite(value):
            raise ValidationError(
                f"{field_name} must be a finite integer: {value} is not finite",
                "SCP-VALID-7005",
            )
        if value != int(value):
            raise ValidationError(
                f"{field_name} must be an integer, got non-integer float {value!r}",
                "SCP-VALID-7005",
            )
    result = int(value)  # type: ignore[arg-type]
    if result < 0:
        raise ValidationError(
            f"{field_name} must be non-negative, got {result}",
            "SCP-VALID-7005",
        )
    return result


# ---------------------------------------------------------------------------
# RevocationStatus
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class RevocationStatus:
    """Revocation status for an identity attestation (§3.5).

    Mirrors the Rust ``RevocationStatus`` enum:

    - ``Active`` -> ``RevocationStatus(status="active")``
    - ``Revoked { revoked_at, reason }`` ->
      ``RevocationStatus(status="revoked", revoked_at=..., reason=...)``
    """

    #: Status string: ``"active"`` or ``"revoked"``.
    status: str

    #: Unix timestamp (seconds) when the attestation was revoked.
    #: Only present when ``status == "revoked"``.
    revoked_at: int | None = None

    #: Optional human-readable revocation reason.
    #: Only present when ``status == "revoked"``.
    reason: str | None = None

    def __post_init__(self) -> None:
        if self.status == "revoked" and self.revoked_at is None:
            raise ValueError("RevocationStatus with status='revoked' requires revoked_at")
        if self.status not in ("active", "revoked"):
            raise ValueError(
                f"Invalid revocation status: {self.status!r} (expected 'active' or 'revoked')"
            )
        if self.revoked_at is not None:
            object.__setattr__(self, "revoked_at", _parse_finite_int(self.revoked_at, "revoked_at"))


# ---------------------------------------------------------------------------
# IdentityAttestation
# ---------------------------------------------------------------------------


@dataclass
class IdentityAttestation:
    """An identity link attestation binding a DID to an external platform (§3.5).

    Pure data class. Creation, listing, removal, renewal, and verification
    live on :class:`scp_sdk.SCP` as
    :meth:`~scp_sdk.SCP.create_identity_link_attestation`,
    :meth:`~scp_sdk.SCP.identity_link_attestations`,
    :meth:`~scp_sdk.SCP.remove_identity_link_attestation`,
    :meth:`~scp_sdk.SCP.identity_renew_attestation`, and
    :meth:`~scp_sdk.SCP.verify_identity_link_attestation`.

    The ``id`` is deterministically derived as
    ``hex(SHA-256(issuer || platform || handle || issued_at))``.
    """

    #: Deterministic attestation ID.
    id: str

    #: Platform identifier (e.g. ``"github.com"``).
    platform: str

    #: Platform handle or username.
    platform_handle: str

    #: DID verification method that signed this attestation
    #: (e.g. ``"did:dht:z6Mk...#active"``).
    verification_method: str

    #: Unix timestamp (seconds) when the evidence was last verified.
    verified_at: int

    #: Revocation status (``"active"`` or ``"revoked"`` with metadata).
    revocation_status: RevocationStatus = field(
        default_factory=lambda: RevocationStatus(status="active")
    )

    #: Optional platform-assigned unique identifier.
    platform_id: str | None = None

    def __post_init__(self) -> None:
        self.verified_at = _parse_finite_int(self.verified_at, "verified_at")

    def _to_bridge_dict(self) -> dict[str, Any]:
        """Convert to a dict for bridge serialization."""
        rs = self.revocation_status
        if rs.status == "revoked":
            rs_value: dict[str, Any] = {"Revoked": {}}
            if rs.revoked_at is not None:
                rs_value["Revoked"]["revoked_at"] = rs.revoked_at
            if rs.reason is not None:
                rs_value["Revoked"]["reason"] = rs.reason
        else:
            rs_value = "Active"  # type: ignore[assignment]

        d: dict[str, Any] = {
            "id": self.id,
            "platform": self.platform,
            "platform_handle": self.platform_handle,
            "verification_method": self.verification_method,
            "verified_at": self.verified_at,
            "revocation_status": rs_value,
        }
        if self.platform_id is not None:
            d["platform_id"] = self.platform_id
        return d

    @classmethod
    def _from_dict(cls, data: dict[str, Any]) -> IdentityAttestation:
        """Construct from a dict returned by the bridge."""
        raw_rs = data.get("revocation_status", "active")
        if isinstance(raw_rs, dict) and "Revoked" in raw_rs:
            revoked_data = raw_rs["Revoked"]
            revoked_at_raw = revoked_data.get("revoked_at")
            if revoked_at_raw is None:
                raise ValueError("Bridge returned Revoked status without revoked_at timestamp")
            rs = RevocationStatus(
                status="revoked",
                revoked_at=_parse_finite_int(revoked_at_raw, "revoked_at"),
                reason=revoked_data.get("reason"),
            )
        elif isinstance(raw_rs, str) and raw_rs.lower() == "active":
            rs = RevocationStatus(status="active")
        elif isinstance(raw_rs, str) and raw_rs.lower() == "revoked":
            raise ValueError("Bridge returned bare 'revoked' string without revocation metadata")
        else:
            raise ValueError(f"Unknown revocation status from bridge: {raw_rs!r}")

        return cls(
            id=data["id"],
            platform=data["platform"],
            platform_handle=data["platform_handle"],
            verification_method=data["verification_method"],
            verified_at=_parse_finite_int(data["verified_at"], "verified_at"),
            revocation_status=rs,
            platform_id=data.get("platform_id"),
        )

    def __repr__(self) -> str:
        return (
            f"IdentityAttestation(id={self.id!r}, platform={self.platform!r}, "
            f"handle={self.platform_handle!r}, status={self.revocation_status.status!r})"
        )


__all__ = [
    "DIDDocument",
    "Identity",
    "IdentityAttestation",
    "RevocationStatus",
]

File: bindings/typescript/src/identity.ts
/**
 * Identity module for the SCP TypeScript SDK.
 *
 * After Phase 4 PR 4 (#1549, ADR-048) Agent B1, the `Identity`,
 * `IdentityAttestation`, and `RevocationStatus` types collapse to pure
 * handle/value types: no `#scp` backing, no instance methods that
 * touch the bridge, no static factories. All lifecycle and attestation
 * operations live on the {@link SCP} class.
 *
 * See ADR-022 in `.docs/adrs/phase-4.md`, ADR-048, and
 * `.docs/scaffold/typescript.md`.
 */

import { ValidationError } from "./errors";
import type { BridgeIdentityHandle } from "./internal/bridge";
import type { SCP } from "./scp";

// ---------------------------------------------------------------------------
// IdentityLinkAttestation
// ---------------------------------------------------------------------------

/** An identity link attestation proving ownership of an external platform identity (§3.5.1). */
export interface IdentityLinkAttestation {
  /** Deterministic attestation ID. */
  id: string;
  /** Always `"identity_link"`. */
  type: string;
  /** The DID that issued this attestation. */
  issuer: string;
  /** Same as issuer for self-attestations. */
  subject: string;
  /** Unix timestamp (seconds) when created. */
  issued_at: number;
  /** Platform identity claim. */
  claim: {
    platform: string;
    platform_handle: string;
    platform_id?: string;
    link_type: string;
  };
  /** Evidence supporting the claim. */
  evidence: {
    method: string;
    /** Opaque proof string per spec §3.5.2 — verifiers MUST use as-is in signature scope. */
    proof: string;
    /** Unix timestamp (integer seconds) when the claim was verified. */
    verified_at: number;
    verifier_did?: string;
  };
  /** Revocation status: `"Active"` or `{ Revoked: { revoked_at, reason, revoked_by } }`. */
  revocation_status:
    | "Active"
    | {
        Revoked: {
          /** Unix timestamp (integer seconds) when the attestation was revoked. */
          revoked_at: number;
          reason: string;
          revoked_by: string;
        };
      };
  /** Ed25519 signature bytes. */
  signature: number[];
}

// ---------------------------------------------------------------------------
// CustodyType
// ---------------------------------------------------------------------------

/** Supported custody methods for identity key management. */
export type CustodyType = "platform" | "in_memory" | "software";

// ---------------------------------------------------------------------------
// Identity
// ---------------------------------------------------------------------------

/**
 * An opaque handle to an SCP identity.
 *
 * `Identity` is a pure data/handle type: it carries the DID string, the
 * custody type, and the raw bridge handle. All lifecycle operations
 * (`create`, `load`, `resolve`, `rotateKey`, agent-key management,
 * migrate, attestation CRUD, recovery, custody migration) live as
 * methods on the {@link SCP} class. Pass an `Identity` as an argument
 * wherever the underlying bridge call needs access to key material.
 *
 * ```typescript
 * const identity = await scp.identityCreate("in_memory");
 * console.log(identity.did); // "did:dht:z6Mk..."
 * await scp.identityRotateKey(identity);
 * ```
 */
export class Identity {
  /** The DID string for this identity (e.g., `"did:dht:z6Mk..."`). */
  readonly did: string;

  /** The custody type used at identity creation. */
  readonly custodyType: string;

  /** @internal Opaque bridge handle — not part of the public API. */
  readonly _rawHandle: BridgeIdentityHandle;

  private constructor(did: string, custodyType: string, rawHandle: BridgeIdentityHandle) {
    this.did = did;
    this.custodyType = custodyType;
    this._rawHandle = rawHandle;
  }

  /**
   * The serialized `scp_identity::DidRotationEvent` produced by
   * {@link SCP.identityMigrate}. Present only on identities returned by
   * `identityMigrate`; `undefined` for all other operations.
   *
   * Callers **MUST** distribute this rotation event to all active context
   * members per spec §3.2.1 step 4b so peers can update their routing
   * tables.
   */
  get rotationEventJson(): string | undefined {
    return this._rawHandle.rotationEventJson;
  }

  /**
   * Constructs an `Identity` from a raw native NAPI handle.
   *
   * The native addon returns an opaque class instance with at least
   * `did` and `custodyType` fields; this helper narrows `unknown` into
   * an `Identity` wrapper. The `scp` parameter is retained for API
   * symmetry with the other `_fromHandle` statics — the handle itself
   * is self-contained so no `SCP` reference is stored.
   *
   * @internal Phase 4 PR 4 (#1549, ADR-048) — used by `SCP` method
   *   forwarders that return identity-typed results.
   */
  static _fromHandle(_scp: SCP, raw: unknown): Identity {
    const handle = raw as BridgeIdentityHandle;
    return new Identity(handle.did, handle.custodyType, handle);
  }
}

// ---------------------------------------------------------------------------
// RevocationStatus
// ---------------------------------------------------------------------------

/**
 * Revocation status for an identity attestation (§3.5).
 *
 * Mirrors the Rust `RevocationStatus` enum:
 *
 * - `Active` -> `RevocationStatus({ status: "active" })`
 * - `Revoked { revoked_at, reason }` ->
 *   `RevocationStatus({ status: "revoked", revokedAt: ..., reason: ... })`
 */
export type RevocationStatusData =
  | { readonly status: "active" }
  | { readonly status: "revoked"; readonly revokedAt: number; readonly reason?: string };

/**
 * Revocation status for an identity attestation (§3.5).
 *
 * Immutable value object. Use `RevocationStatus.active()` or
 * `RevocationStatus.revoked(revokedAt, reason?)` factory methods.
 */
export class RevocationStatus {
  /** Status string: `"active"` or `"revoked"`. */
  readonly status: "active" | "revoked";

  /**
   * Unix timestamp (seconds) when the attestation was revoked.
   * Only present when `status === "revoked"`.
   */
  readonly revokedAt: number | undefined;

  /**
   * Optional human-readable revocation reason.
   * Only present when `status === "revoked"`.
   */
  readonly reason: string | undefined;

  private constructor(data: RevocationStatusData) {
    this.status = data.status;
    if (data.status === "revoked") {
      this.revokedAt = data.revokedAt;
      this.reason = data.reason;
    } else {
      this.revokedAt = undefined;
      this.reason = undefined;
    }
  }

  /** Creates an active revocation status. */
  static active(): RevocationStatus {
    return new RevocationStatus({ status: "active" });
  }

  /** Creates a revoked revocation status. */
  static revoked(revokedAt: number, reason?: string): RevocationStatus {
    if (!Number.isInteger(revokedAt) || revokedAt < 0) {
      throw new ValidationError("revoked_at must be a non-negative integer", "SCP-VALID-7005");
    }
    const truncated = revokedAt;
    const data: RevocationStatusData =
      reason !== undefined
        ? { status: "revoked", revokedAt: truncated, reason }
        : { status: "revoked", revokedAt: truncated };
    return new RevocationStatus(data);
  }

  /** @internal Convert to bridge-compatible representation. */
  _toBridgeValue(): unknown {
    if (this.status === "revoked") {
      const revoked: Record<string, unknown> = {};
      if (this.revokedAt !== undefined) {
        revoked.revoked_at = this.revokedAt;
      }
      if (this.reason !== undefined) {
        revoked.reason = this.reason;
      }
      return { Revoked: revoked };
    }
    return "Active";
  }

  /** @internal Parse from bridge JSON representation. */
  static _fromBridgeValue(raw: unknown): RevocationStatus {
    if (typeof raw === "object" && raw !== null && "Revoked" in raw) {
      const revoked = (raw as Record<string, Record<string, unknown>>).Revoked;
      const revokedAtRaw = revoked?.revoked_at as number | undefined;
      if (revokedAtRaw === undefined) {
        throw new Error("Bridge returned Revoked status without revoked_at timestamp");
      }
      if (!Number.isInteger(revokedAtRaw) || revokedAtRaw < 0) {
        throw new ValidationError("revoked_at must be a non-negative integer", "SCP-VALID-7005");
      }
      return RevocationStatus.revoked(revokedAtRaw, revoked?.reason as string | undefined);
    }
    if (typeof raw === "string") {
      const lower = raw.toLowerCase();
      if (lower === "active") {
        return RevocationStatus.active();
      }
      if (lower === "revoked") {
        throw new Error("Bridge returned bare 'revoked' string without revocation metadata");
      }
    }
    throw new Error(`Unknown revocation status from bridge: ${String(raw)}`);
  }
}

// ---------------------------------------------------------------------------
// IdentityAttestation
// ---------------------------------------------------------------------------

/** Data for an identity link attestation (§3.5). */
export interface IdentityAttestationData {
  /** Deterministic attestation ID. */
  readonly id: string;
  /** Platform identifier (e.g., `"github.com"`). */
  readonly platform: string;
  /** Platform handle or username. */
  readonly platformHandle: string;
  /** DID verification method that signed this attestation. */
  readonly verificationMethod: string;
  /** Unix timestamp (seconds) when the evidence was last verified. */
  readonly verifiedAt: number;
  /** Revocation status. */
  readonly revocationStatus: RevocationStatus;
  /** Optional platform-assigned unique identifier. */
  readonly platformId?: string | undefined;
}

/**
 * An identity link attestation binding a DID to an external platform (§3.5).
 *
 * Pure data/value class after Phase 4 PR 4 Agent B1 — no `#scp`
 * backing, no `verify()` method. Call
 * `scp.identityVerifyLinkAttestation(...)` with the raw attestation
 * JSON when signature verification is needed.
 */
export class IdentityAttestation implements IdentityAttestationData {
  /** Deterministic attestation ID. */
  readonly id: string;
  /** Platform identifier (e.g., `"github.com"`). */
  readonly platform: string;
  /** Platform handle or username. */
  readonly platformHandle: string;
  /** DID verification method that signed this attestation. */
  readonly verificationMethod: string;
  /** Unix timestamp (seconds) when the evidence was last verified. */
  readonly verifiedAt: number;
  /** Revocation status. */
  readonly revocationStatus: RevocationStatus;
  /** Optional platform-assigned unique identifier. */
  readonly platformId?: string | undefined;

  /** @internal Raw JSON string from the bridge for roundtrip verification. */
  readonly _rawJson?: string | undefined;

  constructor(data: IdentityAttestationData, rawJson?: string) {
    if (!Number.isInteger(data.verifiedAt) || data.verifiedAt < 0) {
      throw new ValidationError("verified_at must be a non-negative integer", "SCP-VALID-7005");
    }
    this.id = data.id;
    this.platform = data.platform;
    this.platformHandle = data.platformHandle;
    this.verificationMethod = data.verificationMethod;
    this.verifiedAt = data.verifiedAt;
    this.revocationStatus = data.revocationStatus;
    this.platformId = data.platformId;
    this._rawJson = rawJson;
  }

  /** @internal */
  _toBridgeRecord(): Record<string, unknown> {
    const rec: Record<string, unknown> = {
      id: this.id,
      platform: this.platform,
      platform_handle: this.platformHandle,
      verification_method: this.verificationMethod,
      verified_at: this.verifiedAt,
      revocation_status: this.revocationStatus._toBridgeValue(),
    };
    if (this.platformId !== undefined) {
      rec.platform_id = this.platformId;
    }
    return rec;
  }

  /** @internal */
  static _fromJson(json: string): IdentityAttestation {
    const data = JSON.parse(json) as Record<string, unknown>;
    return IdentityAttestation._fromRecord(data, json);
  }

  /** @internal */
  static _fromRecord(data: Record<string, unknown>, rawJson?: string): IdentityAttestation {
    // Read from nested `claim` and `evidence` structures when present
    // (full attestation JSON), with fallback to flat keys.
    const claim = (data.claim ?? {}) as Record<string, unknown>;
    const evidence = (data.evidence ?? {}) as Record<string, unknown>;

    const platform = (claim.platform ?? data.platform) as string;
    const platformHandle = (claim.platform_handle ??
      data.platform_handle ??
      data.platformHandle) as string;
    const platformId = (claim.platform_id ?? data.platform_id ?? data.platformId) as
      | string
      | undefined;
    const verificationMethod = (evidence.method ??
      data.verification_method ??
      data.verificationMethod) as string;
    const verifiedAtRaw = (evidence.verified_at ?? data.verified_at ?? data.verifiedAt) as number;
    if (!Number.isInteger(verifiedAtRaw) || verifiedAtRaw < 0) {
      throw new ValidationError("verified_at must be a non-negative integer", "SCP-VALID-7005");
    }
    const verifiedAt = verifiedAtRaw;
    const rawRs = data.revocation_status ?? data.revocationStatus ?? "active";
    const revocationStatus =
      rawRs instanceof RevocationStatus ? rawRs : RevocationStatus._fromBridgeValue(rawRs);

    return new IdentityAttestation(
      {
        id: data.id as string,
        platform,
        platformHandle,
        verificationMethod,
        verifiedAt,
        revocationStatus,
        platformId,
      },
      rawJson,
    );
  }
}
